component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/handler/DPoPAuthenticationHandler.java

```diff
@@ -0,0 +1,343 @@
+            }
+        }
+        authenticationResult.setAuthenticationStatus(AuthenticationStatus.SUCCESS);
+        return authenticationResult;
+    }
+
+    /**
+     * Validate access token binding value.
+     *
+     * @param messageContext message context.
+     * @param tokenBinding   token binding.
+     * @param clientId       OAuth2 client id.
+     * @param accessToken    Bearer token from request.
+     * @return true if token binding is valid.
+     */
+    private boolean isTokenBindingValid(MessageContext messageContext, TokenBinding tokenBinding, String clientId,
+                                        String accessToken) {
+
+        if (tokenBinding == null || StringUtils.isBlank(tokenBinding.getBindingReference())) {
+            return true;
+        }
+
+        Request authenticationRequest =
+                ((AuthenticationContext) messageContext).getAuthenticationRequest().getRequest();
+        OAuthAppDO oAuthAppDO;
+        try {
+            oAuthAppDO = OAuth2Util.getAppInformationByClientId(clientId);
+        } catch (IdentityOAuth2Exception | InvalidOAuthClientException e) {
+            log.error("Failed to retrieve application information by client id: " + clientId, e);
+            return false;
+        }
+
+        if (!oAuthAppDO.isTokenBindingValidationEnabled()) {
+            if (authenticationRequest.getRequestURI().toLowerCase().endsWith(DPoPConstants.SCIM2_ME_ENDPOINT_URI) &&
+                    isSSOSessionBasedTokenBinding(tokenBinding.getBindingType())) {
+                setCurrentSessionIdThreadLocal(getTokenBindingValueFromAccessToken(accessToken));
+            }
+            return true;
+        }
+
+        if (OAuth2Util.isValidTokenBinding(tokenBinding, authenticationRequest)) {
+            if (authenticationRequest.getRequestURI().toLowerCase().endsWith(DPoPConstants.SCIM2_ME_ENDPOINT_URI) &&
+                    isSSOSessionBasedTokenBinding(tokenBinding.getBindingType())) {
+                setCurrentSessionIdThreadLocal(tokenBinding.getBindingValue());
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Get the token binding value which corresponds to the current session identifier from the token when
+     * SSO-session-based token binding is enabled.
+     *
+     * @param accessToken Bearer token from request.
+     * @return Token binding value.
+     */
+    private String getTokenBindingValueFromAccessToken(String accessToken) {
+
+        String tokenBindingValue = null;
+        try {
+            AccessTokenDO accessTokenDO = OAuth2Util.findAccessToken(accessToken, false);
+            if (accessTokenDO != null) {
+                if (accessTokenDO.getTokenBinding() != null &&
+                        StringUtils.isNotBlank(accessTokenDO.getTokenBinding().getBindingValue()) &&
+                        isSSOSessionBasedTokenBinding(accessTokenDO.getTokenBinding().getBindingType())) {
```

Comment:
For DPoP tokens we always have DPoP Bindings. We don't need isSSOSessionBasedTokenBinding check. We can check for DPoP binding

Reply:
Removed